xds/internal/balancer/clusterresolver/balancerconfigbuilder/type.go

```diff
@@ -0,0 +1,99 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package balancerconfigbuilder
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"google.golang.org/grpc/grpclog"
+)
+
+// DiscoveryMechanismType is the type of discovery mechanism.
+type DiscoveryMechanismType int
+
+const (
+	// DiscoveryMechanismTypeEDS is eds.
+	DiscoveryMechanismTypeEDS DiscoveryMechanismType = iota // `json:EDS`
+	// DiscoveryMechanismTypeLogicalDNS is DNS.
+	DiscoveryMechanismTypeLogicalDNS // `json:LOGICAL_DNS`
+)
+
+// MarshalJSON marshals a DiscoveryMechanismType to a quoted json string.
+//
+// This is necessary to handle enum (as strings) from JSON.
+func (t DiscoveryMechanismType) MarshalJSON() ([]byte, error) {
+	buffer := bytes.NewBufferString(`"`)
+	switch t {
+	case DiscoveryMechanismTypeEDS:
+		buffer.WriteString("EDS")
+	case DiscoveryMechanismTypeLogicalDNS:
+		buffer.WriteString("LOGICAL_DNS")
+	}
+	buffer.WriteString(`"`)
+	return buffer.Bytes(), nil
+}
+
+// UnmarshalJSON unmarshals a quoted json string to the DiscoveryMechanismType.
+func (t *DiscoveryMechanismType) UnmarshalJSON(b []byte) error {
+	var s string
+	err := json.Unmarshal(b, &s)
+	if err != nil {
+		return err
+	}
+	switch s {
+	case "EDS":
+		*t = DiscoveryMechanismTypeEDS
+	case "LOGICAL_DNS":
+		*t = DiscoveryMechanismTypeLogicalDNS
+	default:
+		return fmt.Errorf("unable to unmarshal string %q to type DiscoveryMechanismType", s)
+	}
+	return nil
+}
+
+// DiscoveryMechanism is the discovery mechanism, can be either EDS or DNS.
+//
+// For DNS, the ClientConn target will be used for name resolution.
+//
+// For EDS, if EDSServiceName is not empty, it will be used for watching. If
+// EDSServiceName is empty, Cluster will be used.
+type DiscoveryMechanism struct {
+	// Cluster is the cluster name.
+	Cluster string `json:"cluster,omitempty"`
+	// LoadReportingServerName is the LRS server to send load reports to. If
+	// not present, load reporting will be disabled. If set to the empty string,
+	// load reporting will be sent to the same server that we obtained CDS data
+	// from.
+	LoadReportingServerName *string `json:"lrsLoadReportingServerName,omitempty"`
+	// MaxConcurrentRequests is the maximum number of outstanding requests can
+	// be made to the upstream cluster. Default is 1024.
+	MaxConcurrentRequests *uint32 `json:"maxConcurrentRequests,omitempty"`
+	// Type is the discovery mechanism type.
+	Type DiscoveryMechanismType `json:"type,omitempty"`
+	// EDSServiceName is the EDS service name, as returned in CDS. May be unset
+	// if not specified in CDS. For type EDS only.
+	//
+	// This is used for EDS watch if set. If unset, Cluster is used for EDS
+	// watch.
+	EDSServiceName string `json:"edsServiceName,omitempty"`
+}
+
+var logger = grpclog.Component("xds")
```

Comment:
This is shared by eds and cluster_resolver, so I cannot put either one here.

When we move this back to cluster_resolver package, this will become `grpclog.Component("clusterresolver")` (actually, we will deleted this as a duplicate definition of var `logger`)